### PR TITLE
Add sequencer role to oracle

### PR DESCRIPTION
**Pull request by maurelian**
**Description**

Introduces a new role, so that the OutputOracle now has the following: 

- `owner` which can:
  - delete outputs
  - change the sequencer's address
  - transfer ownership

- `sequencer` which can:
	- append new outputs 	
 	

**Metadata**
- Fixes ENG-2255

### CODE REVIEW

File: packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol [hunk 1]
```diff
@@ -85,6 +95,23 @@ contract L2OutputOracle is Ownable {
...
+        _;
+    }
+
+    /**
+     * @notice Transfers the sequencer role to a new account (`newSequencer`).
+     *         Can only be called by the current owner.
+     */
+    function changeSequencer(address newSequencer) public onlyOwner {
+        require(newSequencer != address(0), "OutputOracle: new sequencer is the zero address");
+        sequencer = newSequencer;
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol:112]: This should emit an event

File: packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol [hunk 2]
```diff
@@ -139,7 +169,7 @@ contract L2OutputOracle is Ownable {
         uint256 _l2BlockNumber,
         bytes32 _l1Blockhash,
         uint256 _l1BlockNumber
-    ) external payable onlyOwner {
+    ) external payable onlySequencer {
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol:172]: We should have a tool that we can mount the key into an image and run to delete outputs. Not a blocker for this PR but logistically if the same key used to submit batches is used to delete outputs we need to be logistically ready. hh tasks require a non significant amount of memory which isn't ideal but it is a standard way to do so, otherside a bash script that uses cast

File: packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol [hunk 3]
```diff
@@ -86,6 +86,13 @@ contract L2OutputOracleTest is L2OutputOracle_Initializer {
         assertEq(oracle.computeL2Timestamp(startingBlockNumber + 96024), startingTimestamp + submissionInterval * 96024);
     }
 
+    /*******************
+     * Ownership tests *
+     *******************/
+
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol:92]: I see these are still todos

File: packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol [hunk 4]
```diff
@@ -48,6 +48,11 @@ contract L2OutputOracle is Ownable {
         uint256 indexed _l2BlockNumber
     );
 
+    /**
+     * @notice Emitted when the sequencer address is changed.
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol:52]: This needs `@param`s in the natspec

**Comment by tynes** [comment]
Generally looks good to me, looks like the go bindings are out of date which is causing the CI to fail

**Comment by maurelian** [comment]
@tynes:
> We could add something to CI that warns on storage slot changes too if we want to

This is a good idea. WDYT of this approach?
1. Create a directory in the bedrock package called `.layouts`
2. Create a file for each relevant contract in that directory (ie. `.layouts/OptimismPortal`)
2. Create simple bash script that runs `forge inspect <ContractName> storage-layout .layouts/<ContractName>` 
3. Fail if the file git diff on the file is non-empty

**Comment by tynes** [comment]
> @tynes:
> 
> > We could add something to CI that warns on storage slot changes too if we want to
> 
> This is a good idea. WDYT of this approach?
> 
> 1. Create a directory in the bedrock package called `.layouts`
> 2. Create a file for each relevant contract in that directory (ie. `.layouts/OptimismPortal`)
> 3. Create simple bash script that runs `forge inspect <ContractName> storage-layout .layouts/<ContractName>`
> 4. Fail if `git diff` on the file is non-empty

See https://github.com/foundry-rs/foundry/issues/2056

**Review by tuxcanfly — COMMENTED** [review]
Running `forge test` I'm seeing a test failure. Tried rebuilding from scratch too. Weird since CI seems to pass.

`[FAIL. Reason: OutputOracle: Blockhash does not match the hash at the expected height.] test_appendWithBlockhashAndHeight() (gas: 20419)
`

**Comment by maurelian** [comment]
Odd, I can't repro that.
@tuxcanfly what if you try `forge test --force` to recompile the contracts?

**Comment by tuxcanfly** [comment]
Same.

**Comment by maurelian** [comment]
oh.. updating forge does indeed break it for me too... :/

File: packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol [hunk 5]
```diff
@@ -214,7 +256,7 @@ contract L2OutputOracleTest is L2OutputOracle_Initializer {
         L2OutputOracle.OutputProposal memory proposalToDelete = oracle.getL2Output(latestBlockNumber);
         L2OutputOracle.OutputProposal memory newLatestOutput = oracle.getL2Output(latestBlockNumber - submissionInterval);
```
Comment (Nicca42) [inline packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol:258]: Delete tests should be split into happy and sad path like the Append tests.

**Review by Nicca42 — CHANGES_REQUESTED** [review]

File: packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol [hunk 6]
```diff
@@ -86,6 +86,48 @@ contract L2OutputOracleTest is L2OutputOracle_Initializer {
         assertEq(oracle.computeL2Timestamp(startingBlockNumber + 96024), startingTimestamp + submissionInterval * 96024);
     }
 
+    /*******************
+     * Ownership tests *
+     *******************/
+
+    event SequencerChanged(address indexed, address indexed);
```
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol:93]: For the sake of testing this is fine.

File: packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol [hunk 7]
```diff
@@ -230,4 +257,19 @@ contract L2OutputOracle is Ownable {
...
     }
+
+    /**
+     * @notice Transfers the sequencer role to a new account (`newSequencer`).
+     *         Can only be called by the current owner.
+     */
+    function changeSequencer(address newSequencer) public onlyOwner {
+        require(newSequencer != address(0), "OutputOracle: new sequencer is the zero address");
+        require(newSequencer != owner(), "OutputOracle: sequencer cannot be same as the owner");
+        require(
```
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol:268]: Yeah... https://github.com/ethereum-optimism/optimism/pull/2810#discussion_r904042397

But I agree, allowing a no-op is fine. I'll remove it.

File: packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol [hunk 8]
```diff
@@ -250,7 +292,7 @@ contract L2OutputOracleTest is L2OutputOracle_Initializer {
         uint256 previousBlockNumber = oracle.latestBlockNumber() - submissionInterval;
         L2OutputOracle.OutputProposal memory proposalToDelete = oracle.getL2Output(previousBlockNumber);
 
-        vm.prank(sequencer);
+        vm.prank(owner);
         vm.expectRevert("OutputOracle: The output root to delete does not match the latest output proposal.");
         oracle.deleteL2Output(proposalToDelete);
     }
```
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol:336]: good catch

File: packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol [hunk 9]
```diff
@@ -85,6 +98,24 @@ contract L2OutputOracle is Ownable {
...
+        _;
+    }
+
+    /**
+     * @notice Transfers the sequencer role to a new account (`newSequencer`).
+     *         Can only be called by the current owner.
+     */
+    function changeSequencer(address newSequencer) public onlyOwner {
+        require(newSequencer != address(0), "OutputOracle: new sequencer is the zero address");
+        emit SequencerChanged(sequencer, newSequencer);
```
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol]: done

File: packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/contracts/test/L2OutputOracle.t.sol]: but I changed it anyways.

**Review by tynes — APPROVED** [review]

**Pull request merged**